easybuild/easyblocks/generic/cargo.py

```diff
@@ -0,0 +1,131 @@
+# from easybuild.tools.filetools import remove_file
+
+CRATESIO_SOURCE = "https://crates.io/api/v1/crates"
+
+
+class Cargo(EasyBlock):
+    """Support for installing Cargo packages (Rust)"""
+
+    @staticmethod
+    def extra_options(extra_vars=None):
+        """Define extra easyconfig parameters specific to Cargo"""
+        extra_vars = EasyBlock.extra_options(extra_vars)
+        extra_vars.update({
+            # 'tests': [True, "Build tests", CUSTOM],
+            'offline': [True, "Build tests", CUSTOM],
+            'lto': [False, "Build with link time optimization", CUSTOM],
+        })
+
+        # if 'source_urls' not in extra_vars:
+        #    extra_vars['source_urls'] = [CRATESIO_SOURCE]
+
+        # extra_vars['download_filename_template'] = '%(name)s/%(version)s/download'
+
+        return extra_vars
+
+    def __init__(self, *args, **kwargs):
+        """Constructor for Simpack easyblock."""
+        super(Cargo, self).__init__(*args, **kwargs)
+        env.setvar('CARGO_HOME', os.path.join(self.builddir, '.cargo'))
+        env.setvar('RUSTC', 'rustc')
+        env.setvar('RUSTDOC', 'rustdoc')
+        env.setvar('RUSTFMT', 'rustfmt')
+        optarch = build_option('optarch')
+        if not optarch:
+            optarch = 'native'
+        env.setvar('RUSTFLAGS', '-C target-cpu=%s' % optarch)
+        env.setvar('RUST_LOG', 'DEBUG')
+        env.setvar('RUST_BACKTRACE', '1')
+
+    def configure_step(self):
+        pass
+
+    def extract_step(self):
+        """Populate all vendored deps with required .cargo-checksum.json"""
+        EasyBlock.extract_step(self)
+        for source in self.src:
+            dirname = source["name"].rsplit('.', maxsplit=2)[0]
+            self.log.info('creating .cargo-checksums.json file for : %s', dirname)
+            chksum = compute_checksum(source['path'], checksum_type='sha256')
+            chkfile = '%s/%s/.cargo-checksum.json' % (self.builddir, dirname)
+            write_file(chkfile, '{"files":{},"package":"%s"}' % chksum)
+
+    def build_step(self):
+        """Build with cargo"""
+        parallel = ''
+        if self.cfg['parallel']:
+            parallel = "-j %s" % self.cfg['parallel']
+
+        tests = ''
+        if self.cfg['tests']:
+            parallel = "--tests"
+
+        offline = ''
+        if self.cfg['offline']:
+            parallel = "--offline"
+
+        profile = 'debug' if self.toolchain.options.get('debug', None) else 'release'
+
+        lto = ''
+        if self.cfg['lto']:
+            parallel = '--config profile.%s.lto=true' % profile
+
+        run_cmd('rustc --print cfg', log_all=True, simple=True)  # for tracking in log file
+        # attempt to circumvent the checksum-check that cargo build does, but it still looks for the checksum json file
+        # remove_file('Cargo.lock')
+        # Can't figure out how to supply this via command line
+        write_file('.cargo/config.toml', '[source.crates-io]\ndirectory=".."', append=True)
```

Comment:
I haven't thought about what even happens if there would be a different source than crates.io